Add alpha core algorithm

prachics opened the pull request:
# Add alpha_core algorithm for node ranking

This PR adds the implementation of the AlphaCore algorithm from the paper "AlphaCore: Data Depth based Core Decomposition" by F. Victor, C.G. Akcora, Y.R. Gel, M. Kantarcioglu (KDD '21).

## Features
- Implements the first centrality measure that directly incorporates node features into the ranking process
- Uses Mahalanobis depth to identify core nodes without requiring feature scaling or normalization
- Automatically falls back to graph topology features when no node attributes are available
- Returns a hierarchical ranking of nodes with their alpha values and batch IDs

## Use cases
- Identifying influential nodes in directed networks using both structural position and node attributes
- Feature-aware hierarchical decomposition of networks
- Core-periphery structure analysis that considers both network structure and node properties

## Implementation notes
- Works with directed graphs and directed multigraphs
- Handles edge cases such as empty graphs and singular covariance matrices
- Includes comprehensive tests

prachics committed:
Add alpha_core algorithm for node ranking

prachics committed:
Add alpha_core algorithm for node ranking

prachics committed:
Fix doctest in AlphaCore example

prachics committed:
Finalize AlphaCore implementation and tests

prachics committed:
Merge branch 'main' into add-alphaCore-algorithm

prachics committed:
Fix AlphaCore implementation: simplified feature extraction

prachics committed:
Update conftest.py to handle numpy and pandas dependencies for alphacore

cakcora commented:
hi, I wanted to add that Prachi is our student and this pull request is in fact the official contribution of our original AlphaCore repo from https://github.com/friedhelmvictor/alphacore

rossbar commented:
As a start, could you provide a little background info on what this algorithm is/does? In what contexts is it useful? In your opinion, are there any similar algorithms in networkx that could help folks understand? For example, how does this relate to centrality measures (could it be considered a centrality measure?)

cakcora commented:
Thank you for the question. AlphaCore is a core decomposition algorithm designed to work on directed and weighted graphs with multiple node features. Its key innovation is using data depth, specifically Mahalanobis depth, as a way to collapse several node features into a single numeric score that determines how central or peripheral a node is within the multivariate feature space. This enables AlphaCore to use the edges and iteratively remove nodes with low depth (interpreted as outliers or less structurally important) and assign them to progressively lower "α-cores", analogous to how $k$-core decomposition works for degree.

**When is it useful?**
AlphaCore is particularly helpful in settings where nodes have multiple properties (e.g., in- and out-strength, temporal activity, local structure) and the network is directed and weighted. This includes blockchain networks (e.g., identifying important accounts in token transfer graphs), transportation systems, and communication or social networks. Traditional $k$-core or centrality algorithms usually only consider a single scalar feature like degree or edge weight; AlphaCore generalizes to multivariate settings without needing manual weighting or normalization.

**Relation to centrality**:
AlphaCore acts as a hybrid between a core decomposition and a centrality measure. By setting the starting threshold $\epsilon_{\text{start}}$ to a low value (such as 0.1), one first removes nodes with high depth (which tend to be lower in centrality), and then performs a decomposition on the remaining nodes. In this sense, it’s similar to first filtering by centrality and then performing a decomposition among “core-like” nodes. If only a single node feature is used, AlphaCore behaves almost identically to a centrality-based ranking.

**Related functions in networkx**:
There is no direct analog of AlphaCore in networkx, but the following methods are relevant:

nx.core_number() implements $k$-core decomposition, which AlphaCore generalizes.

nx.degree_centrality(), nx.in_degree_centrality(), and nx.out_degree_centrality() provide scalar centrality measures, akin to using a single feature in AlphaCore.

nx.pagerank() and nx.betweenness_centrality() are more global measures, whereas AlphaCore leverages local multivariate structure and iteratively refines the node ranks.

So, AlphaCore can be understood as a multivariate, depth-based generalization of $k$-core, with an optional behavior that mimics a centrality-first filtering step.

cakcora commented:
Thank you for the question. AlphaCore is a core decomposition algorithm designed to work on directed and weighted graphs with multiple node features. Its key innovation is using data depth, specifically Mahalanobis depth, as a way to collapse several node features into a single numeric score that determines how central or peripheral a node is within the multivariate feature space. This enables AlphaCore to use the edges and iteratively remove nodes with low depth (interpreted as outliers or less structurally important) and assign them to progressively lower "α-cores", analogous to how k-core decomposition works for degree.

**When is it useful?**
AlphaCore is particularly helpful in settings where nodes have multiple properties (e.g., in- and out-strength, temporal activity, local structure) and the network is directed and weighted. This includes blockchain networks (e.g., identifying important accounts in token transfer graphs), transportation systems, and communication or social networks. Traditional 
k-core or centrality algorithms usually only consider a single scalar feature like degree or edge weight; AlphaCore generalizes to multivariate settings without needing manual weighting or normalization.

**Relation to centrality:**
AlphaCore acts as a hybrid between a core decomposition and a centrality measure. By setting the starting threshold 
ϵ_start to a low value (such as 0.1), one first removes nodes with high depth (which tend to be lower in centrality), and then performs a decomposition on the remaining nodes. In this sense, it’s similar to first filtering by centrality and then performing a decomposition among “core-like” nodes. If only a single node feature is used, AlphaCore behaves almost identically to a centrality-based ranking.

**Related functions in networkx:**
There is no direct analog of AlphaCore in networkx, but the following methods are relevant:

nx.core_number() implements k -core decomposition, which AlphaCore generalizes.

nx.degree_centrality(), nx.in_degree_centrality(), and nx.out_degree_centrality() provide scalar centrality measures, akin to using a single feature in AlphaCore.

nx.pagerank() and nx.betweenness_centrality() are more global measures, whereas AlphaCore leverages local multivariate structure and iteratively refines the node ranks.

So, AlphaCore can be understood as a multivariate, depth-based generalization of 
k-core, with an optional behavior that mimics a centrality-first filtering step.

prachics committed:
Merge branch 'main' of https://github.com/networkx/networkx into add-alphaCore-algorithm

prachics commented:
Hi @dschult  @rossbar ,
Could someone please assign a reviewer to this PR?

Also, a quick note:
The code and tests are ready, but the CI fails because numpy and pandas are not installed with pip install .[test]. These dependencies are included in the [project.optional-dependencies].default section of pyproject.toml, but not in the test extra.

Would you prefer I add them to the test extra, or should the CI be updated to install both extras (e.g., pip install .[default,test])?

Thanks for your guidance!

MridulS reviewed (commented):
Thanks for your PR.
Note: I haven't read the paper yet!

Here I've just skimmed a bit through the code and I think we can use pandas a lot more and make the code a bit more concise. I haven't tested the suggestions, so please feel free to poke holes in them :)

dschult commented:
NetworkX tests environments without any dependencies because some of our users want no dependencies. To allow those tests to pass for code that **does** use dependencies like numpy or pandas you need to:
- move the `import` statements for those packages inside the functions that use them
- add a line to `networkx/conftest.py` in the section for "needs_numpy" and others.
- skip the tests of functions that use the optional packages using the pytest "import or skip" functionality. Using`np = pytest.importorskip("numpy")` instead of `import numpy as np` at the top of a test module/class/function will skip that module/class/function when testing if the package is not available.

(This should really be in our developer docs...)

prachics committed:
Merge branch 'networkx:main' into add-alphaCore-algorithm

prachics committed:
refactor: handle optional dependencies and improved alphacore.py

prachics committed:
Merge remote-tracking branch 'upstream/main' into add-alphaCore-algorithm

prachics committed:
Fix warning message handling in alphacore tests to match actual behavior

rossbar reviewed (commented):
My high-level take after a quick pass is that the implementation generally does not mesh well with the rest of the library. In particular, it strikes me that there is a lot of code dedicated to converting to/munging data frames that could just as (if not more) efficiently be done with built-in data structures like lists and dicts, or numpy arrays directly.

Now I'm not saying that `pandas` *can't* be used, but I think there are disadvantages to doing so in terms of readability, boilerplate, and accessibility... it's worth noting that no other functions in networkx (except for the converters) depend on pandas. Similarly, adding a dataframe dependency makes expands the footprint of the other dataframe libraries (narwhals, arrow, etc.) to networkx code itself.

Additionally, this version of the code already exists over at https://github.com/friedhelmvictor/alphacore - so users have access to this implementation already. It would make more sense to me for the networkx implementation to better fit with the networkx package as a whole.

Anyways, just my high-level thoughts. Again, I don't intend to *block* anything based on a pandas implementation, but I do feel this would fit much better with NX if the pandas dependency were removed from the function.

prachics committed:
Merge remote-tracking branch 'upstream/main' into add-alphaCore-algorithm

prachics committed:
Remove pandas dependency from AlphaCore implementation

prachics commented:
@rossbar Thanks for the feedback! I've removed the pandas dependency and changed the return type to a dict to match NetworkX conventions.

All tests are passing and the implementation now uses only native Python structures. Ready for your review!

rossbar reviewed (commented):
I think there's still a wider discussion to be had about whether this is a good fit for the library. My main concern is: even after reading the docstring, I don't have any sense of what this function is computing and in which circumstances the alpha core measurement might be meaningful/useful.

Focusing purely on the implementation, there seems to be a lot of repeated code as a result of things being called in the main function and again in the helper functions. E.g. `_get_feature_matrix` is called on L91 in the main function scope, but then is called again inside the scope of `_calculate_mahal_from_center` on L95. IMO the maintenance of multiple copies derived from the same functionality but within different scopes a ripe source of bugs! If there were a way to reduce the indirection or at least be more careful/explicit about scope boundaries I think that would be an improvement.

A couple other high-level questions:
 1. Why convert the numeric features to float? The "default" features (i.e. degrees) will often be integers.
 2. The "default" features include both weighted and unweighted degrees. For unweighted graphs (which is generally a common case) these will be the same. Is there value in duplicating these features then? Does the duplication of degree lead to drastically different results than if the duplicate values were dropped?

See also inline comments!

prachics committed:
Fix AlphaCore: improve docs, fix duplication, remove unsafe assumptions

prachics committed:
Merge branch 'main' into add-alphaCore-algorithm

prachics commented:
@rossbar  Thanks for the detailed review! I've addressed all your concerns:

Documentation: Added clear explanation of when to use AlphaCore (social networks with activity scores, financial networks with risk measures, etc.) and what it computes.

Code duplication: Fixed the repeated _get_feature_matrix calls by computing the matrix and covariance once in the main function.

Unsafe assumption: Removed the assumption that all nodes have the same attributes - now properly handles nodes with different attribute sets.

Float conversion: Added explicit float conversion with comment explaining it's for numerical stability in covariance calculations.
Degree duplication: Only include strength features when edge weights actually vary, avoiding unnecessary duplication.

All tests still pass and the implementation is now more robust. Let me know if you see any other issues!

rossbar commented on networkx/algorithms/alphacore.py:
```diff
@@ -0,0 +1,364 @@
+
+    # Compute covariance matrix for depth calculations
+    matrix = _get_feature_matrix(data)
+    cov = np.cov(matrix.T)
```
The covariance matrix is computed once for the "full" graph (i.e. before any node removals) here... (continue to comment on L138)

rossbar commented on networkx/algorithms/alphacore.py:
```diff
@@ -0,0 +1,364 @@
+            # Recompute features and depth for remaining nodes
+            data = _extract_features(graph, features)
+            matrix = _get_feature_matrix(data)
+            mahal_values = _calculate_mahal_from_center(data, 0, cov, matrix)
```
... then passed into the mahal distance calculation here unmodified. IIUC, this means that the covariance matrix for the "full" graph is used at every iteration in the distance calculation, even after nodes are removed in the above loops. Is this intentional?

cakcora replied:
yes, the covariance matrix is computed only once.

rossbar replied:
@cakcora - this is the latest comment from my review - sorry if it was lost in the noise. I'm still trying to understand the algorithm. If you could provide insight for this test case that would be helpful!

rossbar reviewed (commented):
Thanks for the updates! Looking at the updated code however my opinion hasn't changed much - there's still a lot of indirection and scope crossing that makes the underlying algorithm hard (for me at least) to follow. See e.g. the inline comment related to `cov`.

Also, I see e.g. that the degree/strength has changed as you noted - but I'd still like to know whether or not it mattered. For example, do you expect to get different answers with the duplicated strength (in the unweighted case) or is this just an efficiency concern? If the former, what is the correct behavior?

The reason I'm asking all these questions and pushing so hard for code clarity is that by adding `alpha_core` to networkx, the NetworkX community would be accepting responsibility to maintain this functionality in the future. Being able to understand the algorithm and the moving parts of the implementation are a key component of this promise!

cakcora commented:
Dear @rossbar could you please have a look at the latest comments and let us know if anything is not clear?
